node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1149 @@
+				assignment: ParaId::default(),
+
+				issued_validity: HashSet::default(),
+				backed_candidates: HashSet::default(),
+				seconded: None,
+				backing_watchers: Vec::new(),
+				reported_misbehavior_for: HashSet::default(),
+
+				table: Table::default(),
+				table_context: TableContext::default(),
+			},
+			rx_from,
+		)
+	}
+
+	/// Hand out the sending part of the channel to communicate with this job.
+	fn tx(&mut self) -> mpsc::Sender<ToJob> {
+		self.tx_to.clone()
+	}
+
+	/// Request a validator set from the `RuntimeApi`.
+	async fn request_validators(&mut self) -> Result<Vec<ValidatorId>, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::Validators(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request the scheduler roster from `RuntimeApi`.
+	async fn request_validator_groups(&mut self) -> Result<SchedulerRoster, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidatorGroups(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request the validation code for some `ParaId` at given block
+	/// from `RuntimeApi`.
+	async fn request_validation_code(
+		&mut self,
+		id: ParaId,
+		parent: BlockNumber,
+		parablock: Option<BlockNumber>,
+	) -> Result<ValidationCode, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidationCode(id, parent, parablock, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request a `SigningContext` from the `RuntimeApi`.
+	async fn request_signing_context(&mut self) -> Result<SigningContext, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::SigningContext(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request `HeadData` for some `ParaId` from `RuntimeApi`.
+	async fn request_head_data(&mut self, id: ParaId) -> Result<HeadData, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::HeadData(id, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Logic that runs on startup.
+	async fn on_startup(&mut self) -> Result<(), Error> {
+		let validators = self.request_validators().await?;
+	
+		let roster = self.request_validator_groups().await?; 
+
+		self.table_context.key = signing_key(&validators[..], &self.keystore);
+
+		for assignment in roster.scheduled {
+			self.table_context.groups.insert(
+				assignment.para_id,
+				roster.validator_groups[assignment.group_idx.0 as usize]
+				.iter()
+				.map(|idx| validators[*idx as usize].clone())
```

Comment:
Here we are converting `ValidatorIndex -> ValidatorPublic`. Inside the table context, we are also doing a similar conversion. Why not just have the table context store `HashSet<ValidatorIndex>`, since it will make the implementation of `is_member_of` easier?